### DISCUSSION TITLE
Vaultwarden does not seem to correctly store the user database for organisations

### DISCUSSION THREAD

**Discussion by stanHK** [dani-garcia/vaultwarden#2923] [Q&A]
Not a very technical report, yet one that is a concern. 

After a power-outage, Vaultwarden restarted with the database entries as per import a week ago. 

I am running Vaultwarden on a Synology, for use with my company's team. 
1. I have installed Vaultwarden
2. Created a user
3. Created an organisation
4. Imported the passwords
5. Invited other users
6. Created some new collections

After the Docker went offline due to the power outage - items 1-4 were available. Item 5 and 6 not.

As I run a copy of VW on another Synology machine, after creating more users in the organisation I haveZipped the whole database, copied it to the other Synology. I was able to log back in to vaultwarden, but found exactly the same issue. All users had gone, and only  1-4 were available, not 5 and 6.

So it seems impossible to backup the database including organisation users. Is this a known issue? Any workarounds? JSON of the passwords only imports organization, collections, and passwords, not users.

**Comment by bokkabonga** [comment]
Please provide the support string that you can generate via the admin page.

**Reply by stanHK** [reply]
I can make an emergency access key, but the public domain I am running this on is locked down by CloudFlare Zero Trust. 

I would need at least your Gmail account to give you access.

**Reply by BlackDex** [reply]
@stanHK not sure where you get it from that someone wants access.
He only asked if you could provide the support string which **you** can generate via the admin interface. First go to `your.domain.tld/admin` login and then click on the diagnostics link at the top. There at the bottom is a button to create a support string. That could help us with troubleshooting.

**Reply by stanHK** [reply]
Thanks, I think the confusion is that I don't see a diagnostics link. Could you screen cap the location? I have gone through all the menus.

Version I am on is: Version 2022.10.0

**Reply by bokkabonga** [reply]
As BlackDex stated you need to login to the admin interface. There it will look like this:

![image](https://user-images.githubusercontent.com/62882454/202139253-35eab08d-c535-4d28-b180-2548f1c52b9b.png)

**Reply by stanHK** [reply]
Here you go. Switching off the Docker to set the token made me lose all changes again. But see if this helps. 

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.26.0
* Web-vault version: v2022.10.0
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: false (X-Forwarded-For)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.35.4
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** ADMIN_TOKEN


```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "****/**.*******",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://***.******.***",
  "domain_origin": "*****://***.******.***",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 5 * * * *",
  "emergency_request_timeout_schedule": "0 5 * * * *",
  "enable_db_wal": true,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "Login",
  "smtp_debug": false,
  "smtp_explicit_tls": null,
  "smtp_from": "*****@******.***",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "****.*****.***",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "*****@******.***",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Reply by bokkabonga** [reply]
If restarting makes you loose all settings you definitely do not have configured a persistent storage

**Comment by BlackDex** [comment]
I find the description a bit difficult to understand.
If you mean that after a power-loss the invited users and collections are not visible anymore, it could be that some changes were not fully written (though i find that unlikely). SQLite works with WAL by default, which means Write Ahead Logging, which writes changes very often, and some items could be in the `.wal` or .`shm` file. If those files are still there, including the database, that should work just ok.

Though, it could be that some items were not fully stored yet, and thus because of the power-loss those got lost.
If you say that you have an other Vaultwarden running, and have those users in there including the collections, and then shutdown vaultwarden before you copied that database, or used the `backup` button in the admin interface, that file should be valid and complete.

Only coping the `db.sqlite3` file while Vaultwarden is active could cause the database to be corrupted, because not everything is written back to the database yet.

An other option which i think could also be an issue here, is that you do not have configured a persistent storage.
Without a persistent storage, all the data will be lost when the container is re-created. That could also match that you say that coping the database from the other environment isn't working, if that persistent storage isn't used.

I can assure you that storing users, organizations, collections, groups is working as is should.

**Reply by stanHK** [reply]
_I find the description a bit difficult to understand.
If you mean that after a power-loss the invited users and collections are not visible anymore, it could be that some changes were not fully written (though i find that unlikely)._ --> This is correct

Actually, I was able to reproduce by installing Vaultwarden on another machine and copying the database of my first instance to the second and logging in. All the passwords and organisation were there, but all the users except the main user was gone. The second instance of Vaultwarden was to check the first as described above, and did not include all the users as the database for users seemed corrupt. 

_Only coping the db.sqlite3 file while Vaultwarden is active could cause the database to be corrupted, because not everything is written back to the database yet._ --> I did not copy the db.sqlite3 for the first (original) instance but did copy it to the second instance on the other machine. However, they were Identical.

_An other option which i think could also be an issue here, is that you do not have configured a persistent storage.
Without a persistent storage, all the data will be lost when the container is re-created. That could also match that you say that coping the database from the other environment isn't working, if that persistent storage isn't used._--> I find this a plausible option, though I have followed the instructions for setup. I use a subfolder in the docker folder, would you recommend me to move the DB out of the docker folder? 

Thank you.

**Reply by BlackDex** [reply]
Well, I can't make more out of this then, the database file was corrupted, and that caused issues.
If the other database was also corrupted, or not fully written, or Vaultwarden wasn't closed before you copied the file etc.. that could cause issues.

I suggest to repair the database, or start over.
Creating users, inviting them to organizations is all working.
Also, coping databases (or renaming them) also works fine for me, since i do that from time to time during some different testing.
